File: implement-shell-tools/ls/my-ls.js
#!/usr/bin/env node

const fs = require("node:fs");
const path = require("node:path");

function main() {
  const args = process.argv.slice(2);

  const {
    showAll,
    onePerLine,
    targetDir,
    unknownFlags
  } = parseArgs(args);

  if (unknownFlags.length > 0) {
    console.error(
      `warning: unknown option(s): ${unknownFlags.join(", ")}`
    );
  }

  let files = readDirectory(targetDir);

  files = applyFilters(files, showAll);
  files = sortFiles(files);

  const output = render(files, onePerLine);

  console.log(output);
}

main();

function parseArgs(args) {
  let showAll = false;
  let onePerLine = false;
  let targetDir = ".";
  const unknownFlags = [];

  const knownFlags = new Set(["-a", "-1"]);

  for (const arg of args) {
    if (knownFlags.has(arg)) {
      if (arg === "-a") showAll = true;
      if (arg === "-1") onePerLine = true;
    } else if (arg.startsWith("-")) {
      unknownFlags.push(arg);
    } else {
      targetDir = arg;
    }
  }

  return { showAll, onePerLine, targetDir, unknownFlags };
}

function readDirectory(dir) {
  return fs.readdirSync(dir);
}

function applyFilters(files, showAll) { 
  if (!showAll) {
    return files.filter(f => !f.startsWith("."));
  }

    return [".", "..", ...files];
  }
function sortFiles(files) {
  return files.sort();
}

function render(files, onePerLine) {
  if (onePerLine) {
    return files.join("\n");
  }

  // default format (simple column simulation)
  return files.join("  ");
}

File: implement-shell-tools/ls/my-ls.py
#!/usr/bin/env python3

import sys
import os



def main():
    args = sys.argv[1:]

    show_all = False
    one_per_line = False
    paths = []

    for a in args:
        if a == "-a":
            show_all = True
        elif a == "-1":
            one_per_line = True    
        else:
            paths.append(a)

    if not paths:
        paths = ["."]

    had_error = False

    for path in paths:
        if list_dir(
            path, 
            show_all=show_all, 
            one_per_line=one_per_line,
        ):
            had_error = True

    if had_error:
       sys.exit(1)
       
       
def list_dir(path, show_all=False, one_per_line=False):
    try:
        entries = os.listdir(path)
    except FileNotFoundError:
        print(f"ls: cannot access '{path}': No such file or directory", 
              file=sys.stderr,
        )
        return True
    
    entries = sorted(entries)
    
    if show_all:
        normal = sorted([e for e in entries if not e.startswith('.')])
        hidden = sorted([e for e in entries if e.startswith('.')])
        
        entries = [".", ".."] + normal + hidden
    else:    
        entries = sorted([e for e in entries if not e.startswith('.')])

    for entry in entries:
        print(entry)

    return False

if __name__ == "__main__":
    main()